src/lib.rs

```diff
@@ -160,176 +174,240 @@ where
+    /// Send a message to this websocket
+    pub async fn send(&mut self, msg: Message) -> Result<(), WsError>
+    where
+        S: AsyncWrite + AsyncRead + Unpin,
+    {
+        let f = SendFuture {
+            stream: self,
+            message: Some(msg),
+        };
+        f.await
     }
 
-    fn close(&mut self) -> Poll<(), WsError> {
-        self.inner.close(None).to_async()
+    /// Close the underlying web socket
+    pub async fn close(&mut self, msg: Option<CloseFrame<'_>>) -> Result<(), WsError>
+    where
+        S: AsyncRead + AsyncWrite + Unpin,
+    {
+        let f = CloseFuture {
+            stream: self,
+            message: Some(msg),
+        };
+        f.await
     }
 }
 
-/// Future returned from client_async() which will resolve
-/// once the connection handshake has finished.
-pub struct ConnectAsync<S: AsyncRead + AsyncWrite> {
-    inner: MidHandshake<ClientHandshake<S>>,
+impl<T> Stream for WebSocketStream<T>
+where
+    T: AsyncRead + AsyncWrite + Unpin,
+    AllowStd<T>: Read + Write,
+{
+    type Item = Result<Message, WsError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        trace!("{}:{} Stream.poll_next", file!(), line!());
+        match futures::ready!(self.with_context(Some(cx), |s| {
+            trace!(
+                "{}:{} Stream.with_context poll_next -> read_message()",
+                file!(),
+                line!()
+            );
+            cvt(s.read_message())
+        })) {
+            Ok(v) => Poll::Ready(Some(Ok(v))),
+            Err(WsError::AlreadyClosed) | Err(WsError::ConnectionClosed) => Poll::Ready(None),
+            Err(e) => Poll::Ready(Some(Err(e))),
+        }
+    }
 }
 
-impl<S: AsyncRead + AsyncWrite> Future for ConnectAsync<S> {
-    type Item = (WebSocketStream<S>, Response);
+impl<T> Sink<Message> for WebSocketStream<T>
+where
+    T: AsyncRead + AsyncWrite + Unpin,
+    AllowStd<T>: Read + Write,
+{
     type Error = WsError;
 
-    fn poll(&mut self) -> Poll<Self::Item, WsError> {
-        match self.inner.poll()? {
-            Async::NotReady => Ok(Async::NotReady),
-            Async::Ready((ws, resp)) => Ok(Async::Ready((WebSocketStream::new(ws), resp))),
-        }
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        (*self).with_context(Some(cx), |s| cvt(s.write_pending()))
     }
-}
 
-/// Future returned from accept_async() which will resolve
-/// once the connection handshake has finished.
-pub struct AcceptAsync<S: AsyncRead + AsyncWrite, C: Callback> {
-    inner: MidHandshake<ServerHandshake<S, C>>,
-}
+    fn start_send(mut self: Pin<&mut Self>, item: Message) -> Result<(), Self::Error> {
+        match (*self).with_context(None, |s| s.write_message(item)) {
+            Ok(()) => Ok(()),
+            Err(::tungstenite::Error::Io(ref err))
+                if err.kind() == std::io::ErrorKind::WouldBlock =>
```

Comment:
I think this is the bug. When the inner stream returns Pending / WouldBlock, it needs to register the async context in order to notify the task. Doing this here will result in lost notifications and deadlock.

Reply:
Send method on Futures::Sink doesn't provide a context, and splitting the stream requires a Sink implementation.

Are we just getting lucky with async-std, or is it doing something behind the scene to handle this?

Reply:
Actually, looking at the Sink code, we should get a registration 
```rust
    if let Some(item) = this.item.take() {
            let mut sink = Pin::new(&mut this.sink);
            match sink.as_mut().poll_ready(cx)? {
                Poll::Ready(()) => sink.as_mut().start_send(item)?,
                Poll::Pending => {
                    this.item = Some(item);
                    return Poll::Pending;
                }
            }
        }

        // we're done sending the item, but want to block on flushing the
        // sink
        ready!(Pin::new(&mut this.sink).poll_flush(cx))?;```

`poll_flush` will call `write_pending` in the sync code, which I think might be blocking the pool due to WouldBlock. Trying to see if there is a way to identify WouldBlock before calling `write_pending`.

Reply:
> Are we just getting lucky with async-std, or is it doing something behind the scene to handle this?

I'm fairly sure that we're just getting lucky due to some detail of `async-std`'s scheduler, or something. The contract of Rust's async model is that when we yield, we have to notify the task via the context. I can't imagine `async-std` not also assuming that contract?

Reply:
When start_send is called, it must simply queue the data, not attempt to use the socket. The actual write to the socket should happen in poll_flush. 

One possible fix would be to always return would block from AllowStd without trying to use the AsyncWrite if no context is set. Then the call to poll_flush will go down the path again and trigger the actual write with the context. 
M

Reply:
> One possible fix would be to always return would block from AllowStd without trying to use the AsyncWrite if no context is set. Then the call to poll_flush will go down the path again and trigger the actual write with the context.

That's actually what is happening already from what I can see [here](https://github.com/snapview/tokio-tungstenite/pull/68/files/a27bd616b46fb0c8aab94b913f1c565b32f8e5ae#diff-eeffc045326f81d4c46c22f225d3df90R49) before calling any on the actual `AsyncWrite`. That `Pending` is then converted to `WouldBlock` immediately [here](https://github.com/snapview/tokio-tungstenite/pull/68/files/a27bd616b46fb0c8aab94b913f1c565b32f8e5ae#diff-eeffc045326f81d4c46c22f225d3df90R101). Which one layer up is then converted to an `Ok(())` in `start_send()` [here](https://github.com/snapview/tokio-tungstenite/pull/68/files/a27bd616b46fb0c8aab94b913f1c565b32f8e5ae#diff-b4aea3e418ccdb71239b96952d9cddb6R316). So then `poll_flush()` will call into `write_pending()`, which then does actual writing, would most likely register for being woken up again later.

What am I missing?

Reply:
> Are we just getting lucky with async-std, or is it doing something behind the scene to handle this?

I believe, it is just a bit of pure luck. Tokio 0.2-alpha.6 and async-std don't deadlock since their speed of sending matches the speed of receiving messages, so `WouldBlock` just never occurs. I am thankful to @carllerche for help!

Yet, I still wonder why Tokio master has this disbalance of sent and received messages.

Reply:
You are correct, this is not the bug. I should have checked the actual implementation instead of stopping at "something looks fishy". I dug in more and I have a new hypothesis. I will comment below as it is not related to this thread.

Reply:
Thanks for looking :)